### PR TITLE
Port spanmetricsprocessor to spanmetricsconnector

### PR DESCRIPTION
**Pull request by djaglowski**
As noted in the `spanmetricsprocessor` readme, it has been intended for quite some time that the component will be migrated to become a connector. Now that the connectors framework is mostly in place, I worked through a preliminary implementation of the component as a connector, starting from a copy/paste of the processor. Overall, I think the implementation and user facing config are both much simpler.

For now I'm looking for feedback on any part of the implementation. I believe we need to wait for the next update from the core collector repo before this component will be manually testable.

### CODE REVIEW

**Comment by djaglowski** [comment]
I'm seeing a noticeable performance gain here too:
```
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector
BenchmarkConnectorConsumeTraces-10    	  483807	      2420 ns/op	    1609 B/op	      39 allocs/op

pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor
BenchmarkProcessorConsumeTraces-10    	   53926	     19852 ns/op	   10555 B/op	     127 allocs/op
```

**Comment by jpkrohling** [comment]
cc @kovrus (and @mapno?)

File: connector/spanmetricsconnector/factory.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	// The stability level of the connector.
+	stability = component.StabilityLevelDevelopment
+)
+
+// NewFactory creates a factory for the spanmetrics connector.
+func NewFactory() connector.Factory {
+	return connector.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		connector.WithTracesToMetrics(createConnector, stability),
```
Comment (kovrus) [inline connector/spanmetricsconnector/factory.go:39]: This component emits both traces and metrics. I think we can go for `connector.WithTracesToMetrics` and `connector.WithTracesToTraces`. I am not sure, but it should be able to drop this lock [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18381/files#diff-4cf1d004574eff4ea9a26ff033740136245993c942742b4e8c6622e522599910R242-R258) after branching the pipeline.

**Comment by djaglowski** [comment]
I wrote up a lengthy response to a comment about having the connector pass traces through it like the processor does. The comment was deleted but I'm posting it here for future reference because I think it lays out a good case against it and I expect this idea will come up again.

---

Whether or not the connector should forward the data is a good question - one I expect we'll see a lot. I've gone back and forth on this myself, but landed on the side of not forwarding.

I see this connector as representative of a class of connectors whose primary purpose is to generate new data. I'll call them "generative" connectors for now. The `count` connector is another in this class. Just so we have another example, suppose a `logalerts` connector generates an "alert" (i.e. a log with semantic guarantees) when it sees a FATAL log record. 

Note that there is a meaningful distinction between the generated data and the consumed data. Unfortunately, in many cases these data sets would be mixed together as we emit both to the same pipeline. This problem is not apparent when looking only at `spanmetrics`, but it is a clear issue for `logalerts` and an issue with `count` as well.

There are lots of good reasons someone would want to keep the data sets separate and I think we should prefer a solution that naturally avoids mixing them.

Fortunately, we have a simple solution - don't embed forwarding into generative connectors. In my opinion, this is better in terms of usability as well as development & maintainability. Ultimately, it is not difficult to explicitly use `forward`.

```yaml
# spanmetrics does forward traces
pipelines:
  traces:
    receivers: [foo]
    exporters: [spanmetrics]
  traces/out:
    receivers: [spanmetrics]
    exporters: [jaeger]
  metrics:
    receivers: [spanmetrics]
    exporters: [prometheus]

# spanmetrics does not forwards traces
pipelines:
  traces:
    receivers: [foo]
    exporters: [forward, spanmetrics]
  traces/out:
    receivers: [foward]
    exporters: [jaeger]
  metrics:
    receivers: [spanmetrics]
    exporters: [prometheus]
```

The alternative would seem to be that connector configurations allow users to declare where to send either the generated data or the consumed data. When I tried to work through the details of how exactly this would work, I found that it results in a very inconsistent experience across generative connectors. (Also just to clarify, I'm in favor of referencing pipelines in a connector config, when it's natural. e.g. routing).

Here's a specific usability problem that comes with the pass-through capability. 

A nice feature about connectors is that you can use their capabilities a la carte. Meaning, a user can connect a traces pipeline to a metrics pipeline, and the connector will only perform the function that consumes traces and emits metrics. 

```yaml
# opt out of forwarding traces
pipelines:
  traces:
    receivers: [foo]
    exporters: [spanmetrics]
# Uncomment to forward traces
#  traces/out:
#   receivers: [spanmetrics]
#   exporters: [jaeger]
  metrics:
    receivers: [spanmetrics]
    exporters: [prometheus]
```

Unfortunately this means that this is also a valid config.

```yaml
# 'spanmetrics' behaves exactly as 'forward'
pipelines:
  traces:
    receivers: [foo]
    exporters: [spanmetrics]
  traces/out:
    receivers: [spanmetrics]
    exporters: [jaeger]
```

The more I look at the pass-through the more reason I find not to have it in these generative connectors.

**Comment by djaglowski** [comment]
Closing in favor of #18535

**Pull request closed without merging**